File: examples/python/newt_example.py
from newt import newt
from newt_batch import newt_batch
from newt_dual import newt_dual
import torch
from tensordual import TensorDual
from tensormatdual import TensorMatDual
# A simple quadratic function
def simple_quad(x):
    return torch.tensor([(x[0]-1.0)**2, (x[1]-2.0)**2])

# The Jacobian of the simple quadratic function
def simple_quad_jac(x):
    return torch.tensor([[2*(x[0]-1.0), 0], [0, 2*(x[1]-2.0)]])


x_root, check = newt(torch.tensor([100.0, 10.0]), simple_quad, simple_quad_jac)
print("Root: ", x_root)
print("Check: ", check)


def batch_quad(x):
    M, D = x.shape
    res = torch.zeros((M, D) , dtype=x.dtype, device=x.device)
    res[:,0] =  (x[:, 0]-1.0)**2
    res[:,1] =  (x[:, 1]-2.0)**2
    return res

# The Jacobian of the simple quadratic function
def batch_quad_jac(x):
    M, D = x.shape
    res = torch.zeros((M, D, D), dtype=x.dtype, device=x.device)
    res[:, 0,  0] = 2*(x[:,0]-1.0)
    res[:, 0, 1] = 0.0
    res[:, 1, 0] = 0.0
    res[:, 1, 1] = 2*(x[:,1]-2.0)
    return res


def dual_quad(x):
    res = TensorDual.zeros_like(x)
    res[0] =  (x[0]-1.0)**2
    res[1] =  (x[1]-2.0)**2
    return res

# The Jacobian of the simple quadratic function
def dual_quad_jac(x):
    #Create an zero TensorMatDual using x as a vector
    res = TensorMatDual.createZeroJacFromVec(x)
    res[:, 0, 0] = 2*(x[0]-1.0)
    res[:, 0, 1]  = 0.0
    res[:, 1, 0]  = 0.0
    res[:, 1, 1]  = 2*(x[1]-2.0)
    return res

xis=torch.tensor([[200, 30.0], [150., 100.]], dtype=torch.float64)
print(xis[:,0])
x_root, check = newt_batch(xis, batch_quad, batch_quad_jac)

xisd=TensorDual.create(xis)

x_rootd, checkd = newt_dual(xisd, dual_quad, dual_quad_jac)

print("Roots: ", x_root)
print("Checks: ", check)

File: examples/cpp/newt_example.cpp
#include <torch/torch.h>
#include <iostream>
#include <vector>
#include "newt.hpp"
using Slice = torch::indexing::Slice;



torch::Tensor func(const torch::Tensor& x, const torch::Tensor& params) {

    int M = x.size(0);
    auto y = torch::zeros({M, 2}, torch::dtype(torch::kFloat64));
    auto x1 = x.index({Slice(), 0});
    auto x2 = x.index({Slice(), 1});
    y.index_put_({Slice(), 0}, x2.pow(2)+x1.pow(2)-4.0);
    y.index_put_({Slice(), 1}, x2-x1.exp());
    return y;
}

torch::Tensor jac(const torch::Tensor& x, const torch::Tensor& params) {
    int M = x.size(0);
    torch::Tensor jac = torch::zeros({M, 2, 2}, torch::dtype(torch::kFloat64));
    auto x1 = x.index({Slice(), 0});
    auto x2 = x.index({Slice(), 1});
    jac.index_put_({Slice(), 0, 0}, 2*x1);
    jac.index_put_({Slice(), 0, 1}, 2*x2);
    jac.index_put_({Slice(), 1, 0}, -x1.exp());
    jac.index_put_({Slice(), 1, 1}, 1.0);
    return jac;
}



using namespace janus;
int main() {
  int M =1000;
  torch::Tensor x0 = torch::ones({M, 2}, torch::dtype(torch::kFloat64));
  for ( int i=0; i < M; i++) {
    x0.index_put_({i, 0}, 2.0+0.1*i);
    x0.index_put_({i, 1}, 1.0+0.1*i);
  }
  torch::Tensor params = torch::zeros({M, 2}, torch::dtype(torch::kFloat64));
  auto res = newtTe(x0, params, func, jac);
  auto roots = std::get<0>(res);
  auto check = std::get<1>(res);
  std::cout << "Root=" << roots << std::endl;
  std::cout << "Check=" << check << std::endl;
  std::cout << "Error=" << Jfunc(func(roots, params)) << std::endl;
}